internal/json/json_test.go

```diff
@@ -0,0 +1,186 @@
+package json
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/kylelemons/godebug/pretty"
+)
+
+type StructA struct {
+	Name             string
+	ID               int `json:"id"`
+	Meta             *StructB
+	AdditionalFields map[string]interface{}
+}
+
+type StructB struct {
+	Address          string
+	AdditionalFields map[string]interface{}
+}
+
+type StructC struct {
+	Time             time.Time
+	Project          StructD
+	AdditionalFields map[string]interface{}
+}
+
+type StructD struct {
+	Project          string
+	Info             StructE
+	AdditionalFields map[string]interface{}
+}
+
+type StructE struct {
+	Employees        int
+	AdditionalFields map[string]interface{}
+}
+
+func TestUnmarshal(t *testing.T) {
+	now := time.Now()
+	nowJSON, err := now.MarshalJSON()
+	if err != nil {
+		panic(err)
+	}
+
+	tests := []struct {
+		desc string
+		b    []byte
+		got  interface{}
+		want interface{}
+		err  bool
+	}{
+		{
+			desc: "receiver not a pointer",
+			got:  StructA{},
+			b:    []byte(`{"content": "value"}`),
+			err:  true,
+		},
+		{
+			desc: "receiver not a pointer to a struct",
+			got:  new(string),
+			b:    []byte(`{"content": "value"}`),
+			err:  true,
+		},
+		{
+			desc: "AdditionalFields not a map",
+			b:    []byte(`{"content": "value"}`),
+			got: &struct {
+				AdditionalFields string
+			}{},
+			err: true,
+		},
+		{
+			desc: "Success, no json.Unmarshaler types",
+			b: []byte(
+				`
+				{
+					"Name": "John",
+					"id": 3,
+					"Meta": {
+						"Address": "291 Street",
+						"unknown0": 3.2
+					},
+					"unknown0": 10,
+					"unknown1": "hello"
+				}
+				`,
+			),
+			got: &StructA{},
+			want: &StructA{
+				Name: "John",
+				ID:   3,
+				Meta: &StructB{
+					Address: "291 Street",
+					AdditionalFields: map[string]interface{}{
```

Comment:
Very elegant, in MSAL.NET we did this by extracting each known field from the JSON object and keeping whatever remains as "AdditionalFields". While it works, the code is much more verbose and the token caching code has more concerns that it needs.